tool/tctl/common/idp_command.go

```diff
@@ -0,0 +1,268 @@
+`)
+	s.testAttributeMapping.cmd = testAttrMap
+}
+
+func (s *samlIdPCommand) tryRun(ctx context.Context, cmd string, c auth.ClientI) (match bool, err error) {
+	switch cmd {
+	case s.testAttributeMapping.cmd.FullCommand():
+		return true, trace.Wrap(s.testAttributeMapping.run(ctx, c))
+	default:
+		return false, nil
+	}
+}
+
+// testCommand implements the "tctl idp saml test_attribute_mapping" command.
+type testAttributeMapping struct {
+	cmd *kingpin.CmdClause
+
+	serviceProvider string
+	users           []string
+	outFormat       string
+}
+
+func (t *testAttributeMapping) run(ctx context.Context, c auth.ClientI) error {
+	serviceProvider, err := parseSPFile(t.serviceProvider)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	users, err := getUsersFromAPIOrFile(ctx, t.users, c)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(users) == 0 {
+		return trace.BadParameter("users not found in file: %s", t.users)
+	}
+
+	resp, err := c.SAMLIdPClient().TestSAMLIdPAttributeMapping(ctx, &samlidpv1.TestSAMLIdPAttributeMappingRequest{
+		ServiceProvider: &serviceProvider,
+		Users:           users,
+	})
+	if err != nil {
+		if trace.IsNotImplemented(err) {
+			return trace.NotImplemented("the server does not support testing SAML attribute mapping")
+		}
+		return trace.Wrap(err)
+	}
+
+	switch t.outFormat {
+	case teleport.YAML:
+		if err := utils.WriteYAML(os.Stdout, resp.MappedAttributes); err != nil {
+			return trace.Wrap(err)
+		}
+	case teleport.JSON:
+		if err := utils.WriteJSON(os.Stdout, resp.MappedAttributes); err != nil {
+			return trace.Wrap(err)
+		}
+	default:
+		for i, mappedAttribute := range resp.MappedAttributes {
+			table := asciitable.MakeTable([]string{"Attribute Name", "Attribute Value"})
+			if i > 0 {
+				fmt.Println("---")
+			}
+			fmt.Printf("User: %s\n", mappedAttribute.Username)
+			for name, value := range mappedAttribute.MappedValues {
+				table.AddRow([]string{
+					name,
+					strings.Join(value.Values, ", "),
+				})
+			}
+			fmt.Println(table.AsBuffer().String())
+		}
+	}
+
+	return nil
+}
+
+// parseSPFile parses service provider spec from given file.
+func parseSPFile(fileName string) (types.SAMLIdPServiceProviderV1, error) {
+	var u types.SAMLIdPServiceProviderV1
+	var r io.Reader = os.Stdin
+	if fileName != "" {
+		f, err := os.Open(fileName)
+		if err != nil {
+			return u, trace.Wrap(err)
+		}
+		defer f.Close()
+		r = f
+	}
+
+	decoder := kyaml.NewYAMLOrJSONDecoder(r, defaults.LookaheadBufSize)
+	if err := decoder.Decode(&u); err != nil {
+		if errors.Is(err, io.EOF) {
+			return u, trace.BadParameter("service provider not found in file: %s", fileName)
+		}
+		return u, trace.Wrap(err)
+	}
+	return u, nil
+}
+
+// getUsersFromAPIOrFile parses user from spec file. If file is not found, it fetches user from backend.
+func getUsersFromAPIOrFile(ctx context.Context, userfileOrNames []string, c auth.ClientI) ([]*types.UserV2, error) {
+	ufromFileOrName := flattenSlice(userfileOrNames)
```

Comment:
What is the `ufrom` part of this var name?

Reply:
sorry bad naming! it is just a flattened names `["user1,user2", "user3"]` becomes `["user1", "user2", "user3"]`. I have now updated it to be `flattenedUsernamesOrFileNames` which I think is a bit clear.